mirgecom/array_context.py

```diff
@@ -100,6 +110,152 @@ def actx_class_is_numpy(actx_class: Type[ArrayContext]) -> bool:
         return False
 
 
+def _check_cache_dirs_node() -> None:
+    """Check whether multiple ranks share cache directories on the same node."""
+    from mpi4py import MPI
+
+    size = MPI.COMM_WORLD.Get_size()
+
+    if size <= 1:
+        return
+
+    from mirgecom.mpi import shared_split_comm_world
+
+    with shared_split_comm_world() as node_comm:
+        node_rank = node_comm.Get_rank()
+
+        def _check_var(var: str) -> None:
+            from warnings import warn
+
+            try:
+                my_path = os.environ[var]
+            except KeyError:
+                warn(f"Please set the '{var}' variable in your job script to "
+                    "avoid file system overheads when running on large numbers of "
+                    "ranks. See https://mirgecom.readthedocs.io/en/latest/running/large-systems.html "  # noqa: E501
+                    "for more information.")
+                # Create a fake path so there will not be a second warning below.
+                my_path = f"no/such/path/rank{node_rank}"
+
+            all_paths = node_comm.gather(my_path, root=0)
+
+            if node_rank == 0:
+                assert all_paths
+                if len(all_paths) != len(set(all_paths)):
+                    hostname = MPI.Get_processor_name()
+                    dup = [path for path in set(all_paths)
+                                if all_paths.count(path) > 1]
+
+                    from warnings import warn
+                    warn(f"Multiple ranks are sharing '{var}' on node '{hostname}'. "
+                        f"Duplicate '{var}'s: {dup}.")
+
+        _check_var("XDG_CACHE_HOME")
```

Comment:
Strictly speaking, don't we want `XDG_CACHE_HOME` to be unique even when not using pyopencl? Might be worth keeping the `XDG_CACHE_HOME` check inside `mpi_entry_point` and doing the rest here?

Reply:
`XDG_CACHE_HOME` only matters for a mirgecom run if we use pyopencl, as far as I can tell.

Reply:
Yeah, I wasn't sure what all goes in there. On the one hand I do see some loopy output when running with `--numpy` (clean cache):

```
% ls .cache/xdg-scratch/rank0/pytools 
pdict-v4-loopy-code-gen-cache-v3-2024.1-islpy2023.2.5-cgen2020.1-cc795034cc40d49deed34705ca594c96f407a70f-v1-py3.11.8.final.0
pdict-v4-loopy-invoker-cache-v10-2024.1-islpy2023.2.5-cgen2020.1-cc795034cc40d49deed34705ca594c96f407a70f-v1-py3.11.8.final.0
pdict-v4-loopy-memoize-cache-buffer_array-LoopyKeyBuilder.LoopyKeyBuilder-v0-2024.1-islpy2023.2.5-cgen2020.1-cc795034cc40d49deed34705ca594c96f407a70f-v1-py3.11.8.final.0
pdict-v4-loopy-memoize-cache-contract_arrays-LoopyKeyBuilder.LoopyKeyBuilder-v0-2024.1-islpy2023.2.5-cgen2020.1-cc795034cc40d49deed34705ca594c96f407a70f-v1-py3.11.8.final.0
pdict-v4-loopy-memoize-cache-fuse_same_discretization_entity_loops-LoopyKeyBuilder.LoopyKeyBuilder-v0-2024.1-islpy2023.2.5-cgen2020.1-cc795034cc40d49deed34705ca594c96f407a70f-v1-py3.11.8.final.0
pdict-v4-loopy-memoize-cache-preprocess_program-LoopyKeyBuilder.LoopyKeyBuilder-v0-2024.1-islpy2023.2.5-cgen2020.1-cc795034cc40d49deed34705ca594c96f407a70f-v1-py3.11.8.final.0
pdict-v4-loopy-schedule-cache-v4-2024.1-islpy2023.2.5-cgen2020.1-cc795034cc40d49deed34705ca594c96f407a70f-v1-py3.11.8.final.0
pdict-v4-loopy-typed-and-scheduled-cache-v1-2024.1-islpy2023.2.5-cgen2020.1-cc795034cc40d49deed34705ca594c96f407a70f-v1-py3.11.8.final.0
```
but I'm not sure if this should be happening (e.g., why is there fusion contractor stuff?).

Reply:
Which case did you run?

Reply:
combozzle

Reply:
Hmm, any specific options you use? The cache is empty when I run locally.

Reply:
Wait a sec, are there actually any files in the folders you mentioned above? If not your cache is empty, but the directory structure is created by loopy since it is still imported.

Reply:
`find xdg-cache/ -type f` or so

Reply:
Yeah, they're empty.